oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,492 @@
+----------
+
+Each MFE must bundle all its own dependencies at build time, creating significant overhead for the Webpack build process. When building multiple MFEs, this additional, repeated overhead adds up quickly, making builds prohibitively slow.
+
+Dependency Maintenance
+----------------------
+
+Since each MFE has its own complete set of dependencies, the overhead of keeping them all up to date can be overwhelming. Security patches, bug fixes, new features, and breaking changes all add up and create a significant maintenance burden.
+
+Composability
+=============
+
+We have no clean, low-overhead way of composing components from multiple MFEs into a single page, or extending an MFE with additional functionality.
+
+The reality of MFEs is that while each application may represent a cohesive domain or bounded context, sometimes you need content and functionality from more than one domain on the page at the same time.
+
+Alternately, providers may want to show different *versions* of MFEs to different users while keeping the rest of the app (header, navigation, other MFEs) unchanged.
+
+We aren't without options here, but what we have today has its own issues. It's worth walking through what we have today.
+
+Shared Libraries (Current Option #1)
+------------------------------------
+
+Because each MFE is siloed from each other - both in repositories and at runtime - our only viable mechanism for sharing code has historically been to extract it into a shared library that gets bundled with each at build time. This creates more repository/dependency overhead.
+
+Build-time package overrides (Current Option #2)
+------------------------------------------------
+
+NPM and package.json allow operators to override dependency resolution by installing an alternate version of a dependency prior to build time. This has historically been how we've allowed operators to override the header, footer, and brand.
+
+The system is confusing, somewhat brittle, and only works at build time. If an operator needs different headers/footers/brands for different clients, we've just multiplied the number of builds they need to do (already a problem even without this - see the "Build time" issue above!)
+
+Frontend Plugins (Current Option #3)
+------------------------------------
+
+With the creation of the `frontend-plugin-framework <https://github.com/openedx/frontend-plugin-framework>`_, we're starting to have the capability to share components across MFEs as plugins, either at build time (direct plugins) or runtime (iframe plugins)
+
+Direct plugins create some flexibility, but couple our repositories' builds together. Iframes are good for sandboxing and isolating code, but they're a very inefficient way to compose a UI, especially given the consistency and UX/DX concerns raised above. In a way, they exacerbate the problem even more. There's no way to do direct plugins via the MFE config API, since they rely on importing modules directly into the build.
```

Comment:
`couple our repositories' builds together.` -> who is "our" in this case?

and, in example 742 of why I hate soft wraps, I have to ask a completely different question in the same thread - what is DX?

Reply:
> what is DX?

Developer experience. This seems to be the only place in the OEP where UX is used instead of spelling out user experience, might be worth just spelling out both.

Reply:
Ah, I'm used to "DevExp" or "DevX" in our community. Kyle says "DevX"

Reply:
Avoiding the problem by spelling them both out as Brian suggests. 😂